### PR TITLE
Add Cachi2 Logging Standards

### PR DESCRIPTION
**Pull request by fepas**
A better-defined logging format will help users to better understand the status of requests.
INFO messages should be enough to let the user know what is happening with a request.

# Maintainers will complete the following section

- [ ] Commit messages are descriptive enough
- [ ] Code coverage from testing does not decrease and new code is covered
- [ ] New code has type annotations
- [ ] Docs updated (if applicable)
- [ ] Docs links in the code are still valid (if docs were updated)

### CODE REVIEW

**Comment by fepas** [comment]
https://github.com/fepas/cachi2/blob/STONEBLD-451/docs/logs.md <- to read formatted text

File: docs/logs.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
+# Cachi2 Logging Standards
+
+## Logging principles
+
+1) Use a Consistent Format: To ensure that logs are easy to read and understand, use a consistent format throughout the system. This format should include the date and time of the log entry, the severity level of the log message, and the message itself.
+2) Use Descriptive Severity Levels: Use descriptive severity levels to indicate the importance of a log message. For example, use "INFO" for informational messages, "WARN" for problems that do not interfere with the final result of the request, and "ERROR" for errors that affect the final result of the request.
```
Comment (chmeliik) [inline docs/logs.md:6]: Copying Bruno's points from the original design doc

> Should we also detail here the use of DEBUG?

> Maybe we draw inspiration from Syft.
> 
> The INFO level output is very lean, and it just outlines the overall steps Syft is taking:
> <https://pastebin.com/wBuGj5u5>
> 
> The DEBUG level is very detailed:
> <https://pastebin.com/3Pw9dfwc>

Comment (chmeliik) [reply docs/logs.md:6]: DEBUG is interesting. Currently, we don't enable debug logs by default (not even in rhtap) so any info logged at debug level is as good as lost.

Comment (brunoapimentel) [reply docs/logs.md:6]: We could make debug level the default in RHTAP. I think the question here is more of "how verbose we want INFO to be", since we're setting guidelines.

Do we want to go for a default clean approach, or a default verbose approach? I think it's fine if we decide going towards the latter, but then we'd need to set specific guidelines on what should be DEBUG (if anything at all).

File: docs/logs.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
+# Cachi2 Logging Standards
+
+## Logging principles
+
+1) Use a Consistent Format: To ensure that logs are easy to read and understand, use a consistent format throughout the system. This format should include the date and time of the log entry, the severity level of the log message, and the message itself.
```
Comment (brunoapimentel) [inline docs/logs.md:5]: Maybe we don't need to have the principles laid out in more general terms here. For example, I'd make this section something like:

```
1) Log format: 
[timestamp] [LEVEL] [module.path] [message] [json data]
```

I think this doc should be as specific and objective as possible. The redhat-appstudio one is a good reference for this IMO, ours should be even simpler than that.

File: docs/logs.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+2023-03-07T11:32:29.949Z INFO pkg_managers.pip Found metadata {"name": "cachito-pip-with-deps", "version": "1.0.0" }
+2023-03-07T11:32:29.949Z INFO pkg_managers.pip Resolved package {"name": "cachito-pip-with-deps", "version": "1.0.0"}
+2023-03-07T11:32:29.951Z INFO pkg_managers.pip No hash options used, will not require hashes unless HTTP(S) dependencies are present.
+2023-03-07T11:32:29.959Z INFO pkg_managers.pip Dependencies found [{"name": "ruamel-yaml-clib", "version": "0.2.6"}, {"name": "six", "version": "1.16.0"}, {"name": "urllib3", "version": "1.26.7"}, {"name": "cython", "version": "0.29.33"}]
+2023-03-07T11:32:29.161Z INFO pkg_managers.pip Downloading dependencies [{"name": "ruamel-yaml-clib", "version": "0.2.6"}, {"name": "six", "version": "1.16.0"}, {"name": "urllib3", "version": "1.26.7"}, {"name": "cython", "version": "0.29.33"}]
+2023-03-07T11:32:29.161Z INFO pkg_managers.pip Dependencies downloaded successfully [{"name": "ruamel-yaml-clib", "version": "0.2.6"}, {"name": "six", "version": "1.16.0"}, {"name": "urllib3", "version": "1.26.7"}, {"name": "cython", "version": "0.29.33"}]
```
Comment (chmeliik) [inline docs/logs.md:31]: Should we log the entire list of dependencies on a single line? It already gets long with 3 deps, imagine npm requests with 2000 dependencies

And we probably only need to log them once, not 3 times as "found -> downloading -> downloaded"

File: docs/logs.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+## Consequences
+
+* By using structured logs that are only partially formatted as JSON, we should strike a balance
+  between easy readability and support for centralized queries. It should be easy for a human to
+  read Cachi2 logs directly since the first portion of each line contains a human
+  readable string early on, while the JSON formatted suffix supports the creation of queries that
+  span subsystems in centralized aggregated logs.
+
+
+Note: This document is based on https://github.com/redhat-appstudio/book/blob/main/ADR/0006-log-conventions.md.
```
Comment (chmeliik) [inline docs/logs.md:67]: Similarly to the appstudio ADR, should we look into what this means from an implementation perspective?

I wouldn't want to log like this all over the codebase
```python
log.info("Dependencies found %s", json.dumps(dependencies))
```
It's probably OK to do the detailed design in a follow-up story, basically I'm just asking if structured logging is a thing in Python

File: docs/logs.md (hunk 1 above)
Comment (chmeliik) [reply docs/logs.md:6]: I think, more realistically, we should identify the most important events to log. Some of those might be:

* Starting to process a package
* Resolving the package name and version
  * And the project files parsed / commands used to do that
* Identifying dependencies
  * And the lockfiles parsed / commands used to do that
* Validating dependencies
  * maybe also what validation is being performed?
* Downloading dependencies

**Comment by chmeliik** [comment]
Note that the logging ADR is currently being revised: https://github.com/redhat-appstudio/book/pull/77

TLDR, going full JSON. I don't think we should follow that anymore, unlike controller logs, ours are user facing.

Maybe we just stick with basic key=value pairs?

**Comment by slimreaper35** [comment]
@here  How about this old PR?

**Comment by eskultety** [comment]
> @here How about this old PR? 

My take on this is that we don't need an ADR for this. `appstudio` is  a different beast comprising of [many components](https://github.com/redhat-appstudio/book/blob/main/book/index.md#application-context) for which they need to unify the logging format to something which can be at least remotely useful should a problem arise. IMO we only need to make sure that we're at least somewhat compatible with the format appstudio defines provided cachi2 is supposed to be part of any of the controller components.
In general I don't think we need anything more than a single mention in contributor's guidelines with e.g.

`[timestamp] [LEVEL] [module.path] [message]`

since the logging principles in general are kind of obvious.
I deliberately dropped `jsondata` from the example since the problem with always adding JSON data everywhere (e.g. logging hundreds of deps with every procedure) is that the log will just go from readable (originally) to a incomprehensible data dump - the data will be there, but it'll be a nightmare to navigate through and find the actual issue. We should IMO only include the JSON data/key=value pairs in certain checkpoints, not with every single message.

> I think, more realistically, we should identify the most important events to log. Some of those might be:

    Starting to process a package
    Resolving the package name and version
        And the project files parsed / commands used to do that
    Identifying dependencies
        And the lockfiles parsed / commands used to do that
    Validating dependencies
        maybe also what validation is being performed?
    Downloading dependencies
        maybe log each url (probably at DEBUG) for audit purposes?

^This is more important to define so that we don't end up that massive data dump I mentioned.

**Comment by slimreaper35** [comment]
If there is anything to be done within this PR, I can take that over.

**Comment by brunoapimentel** [comment]
I will take the liberty of closing this PR, since I believe we won't be following the direction here. Some clarifications/comments:

> IMO we only need to make sure that we're at least somewhat compatible with the format appstudio defines provided cachi2 is supposed to be part of any of the controller components.

Cachi2 will be executed as part of a Tekton [task](https://github.com/redhat-appstudio/build-definitions/blob/main/task/prefetch-dependencies/0.1/prefetch-dependencies.yaml), which means that its logs will be standard pipelineRun logs, which are actually separated from controller logs.

> This is more important to define so that we don't end up that massive data dump I mentioned.

I completely agree. I think we need guidelines on what to log and on which level. We still have a backlog story to revisit that. I think we should also revisit the overall UX of the CLI.

**Comment by brunoapimentel** [comment]
Closing this since we won't be following the direction laid out here. The work to improve the overall logs experience and guidelines is postponed to when STONEBLD-451 is prioritized.
